theseus/utils/examples/motion_planning/motion_planner.py

```diff
@@ -53,6 +53,57 @@ def _copy_impl(self, new_name: Optional[str] = None) -> "_XYDifference":
         )
 
 
+class Nonholonomic(th.CostFunction):
+    def __init__(
+        self,
+        vel: th.Vector,
+        cost_weight: th.CostWeight,
+        name: Optional[str] = None,
+    ):
+        super().__init__(cost_weight, name=name)
+        if vel.dof() != 3:
+            raise ValueError(
+                "Nonholonomic only accepts 3D vectors, " "representing vx, vy, vtheta"
+            )
+        self.vel = vel
+        self.register_optim_vars(["vel"])
+        self.weight = cost_weight
+
+    def dim(self):
+        return 1
+
+    def _compute_error(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        vel = self.vel.tensor.unsqueeze(2)
+        cos_v_theta = vel[:, 2].cos()
+        sin_v_theta = vel[:, 2].sin()
+        error = vel[:, 1] * cos_v_theta - vel[:, 0] * sin_v_theta
```

Comment:
I think this might not be entirely correct. Given
pos = [x, y, theta] (or SE2) and vel = [vx, vy, vtheta]
We want
error = vy * cos(theta) - vx * sin(theta)

Ref [Slide 6, [here](https://www.cs.cmu.edu/~16311/current/schedule/ppp/Lec18-Non_Holonomic_Constraints.pdf)], and [gpmp2 code](https://github.com/gtrll/gpmp2/blob/main/gpmp2/dynamics/VehicleDynamics.h#L39).

The current version is using vtheta instead of theta; this cost function will also need pose as an input aside from velocity.

Reply:
IIRC, I had problems with the version that uses `theta` instead of `vtheta`, but it's been a while. Also, maybe this change was related to the formulation where we only had poses and not velocities. I'll revisit this with this example.

Reply:
Sounds good on revisiting. Is there a reference for the version that uses vtheta? I am not sure mathematically we can replace theta for vtheta.

Reply:
I don't remember what was my rationale back then. I remember looking at [this](http://robotics.caltech.edu/wiki/images/4/4b/IntroNoholonomicMotionPlanning.pdf) among other things, but I'm looking at it right now and the current code doesn't really make sense lol. 

Back then we were changing a lot both this and the `Dynamics` cost, and adding/removing velocities, so maybe this made sense for some version of the code that no longer exists. Not really sure.

Reply:
Oh lol no worries. Good we can get it fixed now.

Reply:
Fixed now. Implemented in 3010936f70e0e1e2f58899c44703344222356366. Unit tests in 7413a708b16acee8abd9e96b453e25564897bb68.